### ISSUE TITLE
[Epic] Replay: Load replay data on-demand

### ISSUE THREAD

**Issue by bruno-garcia** [getsentry/sentry#66472]
## Problem

tl;dr: The way the Replay Details page loads the data can be problematic for customer using `canvas` recording. This is specially problematic for users that have huge canvas (4k), in replays that are long. 

Also it's a problem for people on slow network connections, and is a blocker for us supporting longer replays.

## Additional context

Currently, the Session Replay's breadcrumbs (including console, network, etc) and rrweb data used in the details page is served up by a single endpoint. To build the event timeline, as well as fast forward a replay to a specific point in time, we need all relevant data for the duration of the replay.

For that reason, basically we're loading all of the replay data (that can be 60 minutes long) all at once.

This approach worked out so far for the most part but has some disadvantages that became more significant with the introduction of `canvas` support. With the current version of the SDK, we record screenshots for the `canvas` element per second (2 fps). This can lead to the replay details page loading up to 7,200 images all at once (on a max length replay of 60 min, that's 3600 seconds X 2 canvas snapshots).

In order to improve the performance of the Replay playback, we need to be able to fetch the `canvas` snapshot on demand, as playback gets close to that point in time. 

Worth noting that the rrweb recordings themselves need to be loaded from the start all the way to the point in time where we need to display (e.g: link to replay at the time of an error). But that's not the case with `canvas` snapshots.

**Comment by cmanallen** [comment]
Context: mobile replays send video events.  These video events are fetched independently of the RRWeb recording data.  However, they're also stored separately.

**Comment by cmanallen** [comment]
Option.  On download call `json.loads(downloaded_data)`, loop through the `[events]`, extract the breadcrumb events and return, or extract the RRWeb events and return.

This would be backwards compatible but expensive.

**Comment by cmanallen** [comment]
Option: On upload, store breadcrumb events and RRWeb events independently of one another.  On request lookup the `{filename}-rrweb` file or the `{filename}-breadcrumbs` file based on the endpoint or query parameter.  Older files would not be split and would either need a backwards compatible fallback (following the option above) or the client would have to fetch the legacy, combined files and split them locally.

**Comment by jas-kas** [comment]
Example crashes - https://github.com/getsentry/team-replay/issues/399

**Comment by bruno-garcia** [comment]
Created a ticket specific to the investigation so we don't need ot commit to this solution: 

* https://github.com/getsentry/sentry/issues/66519

**Comment by cmanallen** [comment]
We ended up packing the video and rrweb data together.  We can adopt a similar approach for breaking out components of a replay.  We could store breadcrumbs contiguously, canvas events contiguously, etc.  Then return the subset we need for the request type.  The client can request all the breadcrumbs but just the first 5 seconds of canvas data.

**Comment by cmanallen** [comment]
Really we don't need to pack at all.  We could just iterate over the events and filter out the garbage we don't care about.  Has performance implications I'm sure but is simple to bootstrap and prove out.

**Comment by cmanallen** [comment]
Closed because the original problem was solved. Can be reopened for feature additions.
